### ISSUE TITLE
[css-selectors-4] Clarification needed for scoped selectors (§3.3 and §8.6)

### ISSUE THREAD

**Issue by danburzo** [w3c/csswg-drafts#5712]
Hi, I'm having a bit of a hard time understanding the terminology in CSS Selectors Level 4 regarding scoped selectors:

* [3.3 Scoped Selectors](https://drafts.csswg.org/selectors/#scoping)
* [8.6 The Reference Element Pseudo-class: `:scope`](https://drafts.csswg.org/selectors/#the-scope-pseudo)

In particular, I find it hard to follow what is meant by _scoping root_ vs. _:scope elements_ in the absence of examples pointing out which is which, in something like `elem.querySelector('...')`.

I'd love to provide more detailed suggestions on the content of the two sections, but I feel I need a mental model of what's meant in them. Could someone explain, in layman's terms, what a _scoping root_ and a _:scope element_ are for the commonplace DOM methods (matches/querySelector/querySelectorAll)?

**Comment by Loirooriol** [comment]
So `elem.querySelector('...')` is defined in https://dom.spec.whatwg.org/#dom-parentnode-queryselector, it runs [scope-match a selectors string](https://dom.spec.whatwg.org/#scope-match-a-selectors-string) `'...'` against `elem`. This basically sets `elem` as the scoping root when matching.

And since the `:scope` elements are not explicitly specified, but the selector is scoped and the scoping root is the element `elem`, then `elem` matches `:scope` (and nothing else matches).

It's basically the same for `querySelectorAll`: https://dom.spec.whatwg.org/#dom-parentnode-queryselectorall

For [`matches`](https://dom.spec.whatwg.org/#dom-element-matches), the element is used as the `:scope` element but not as the scoping root.

**Comment by danburzo** [comment]
@Loirooriol Thank you for the reply! Just to make sure I got that, I'm going to paraphrase your explanation:

The DOM methods `Element.querySelector(s)`/`Element.querySelectorAll(s)` use their Element instance (`this`) as a _scoping root_.

_:scope elements_ is a mechanism for methods invoking the _match a selector against an element_ algorithm to override the meaning of the `:scope` selector, which would otherwise fall back to the _scoping root_, if any (`:root` if there's no scoping root). 

In contrast to `querySelector`/`querySelectorAll`, `matches()` and `closest()` don't scope their matches to `this` (most obvious for `closest`, which needs to reach up the DOM tree), but they do want `:scope` to match `this`, and not `:root`. 

If that's the case, my top suggestions (both of them IMHO):

* In §3.3, the _Note_ needlessly, and prematurely, introduces the concept of `:scope elements`. It does not clarify anything going on in §3.3 and may better be suited in §8.6 (or even made redundant by it). 
* In §8.6:
  > In some contexts, selectors can be matched with an explicit set of :scope elements. This is is a (potentially empty) set of elements that provide a reference point for selectors to match against, __such as that specified by the querySelector() call in [DOM]__.
  had introduced the confusion — the paragraph suggest `elem` in `elem.querySelector` is a _:scope element_, when in fact it's the _scoping root_. The paragraph should probably read `closest()` instead?

**Comment by cdoublev** [comment]
@danburzo, I believe your concerns have been addressed — intentionally or not — as part of f6337dd.

**Comment by danburzo** [comment]
Thanks for the ping, @cdoublev. The sections are indeed much clearer now, and the edits seem to have addressed my initial concerns. From my point of view the issue can be closed.

**Comment by cdoublev** [comment]
No problem. I was also reviewing the sections related to scoped selectors, and a little follow-up effort benefits everyone digging into existing issues.
